libdnf5/rpm/package_query.cpp

```diff
@@ -2828,4 +2828,51 @@ void PackageQuery::filter_installonly() {
     filter_provides(installonly_packages, libdnf5::sack::QueryCmp::GLOB);
 }
 
+static const std::unordered_set<std::string> CORE_PACKAGE_NAMES = {
+    // See https://access.redhat.com/solutions/27943. These packages should all
+    // also have a reboot_suggested advisory, but it's good to handle them
+    // explicitly, just to be sure.
+    "kernel",
+    "kernel-core",
+    "kernel-PAE",
+    "kernel-rt",
+    "kernel-smp",
+    "kernel-xen",
+    "linux-firmware",
+    "dbus",
+    "glibc",
+    "hal",
+    "systemd",
+    "udev",
+    "gnutls",
+    "openssl-libs",
+    "dbus-broker",
+    "dbus-daemon",
+};
+
+void PackageQuery::filter_reboot_suggested() {
+    auto & pool = get_rpm_pool(p_impl->base);
+    libdnf5::solv::SolvMap filter_result{pool.get_nsolvables()};
+
+    for (const auto & pkg : *this) {
+        if (CORE_PACKAGE_NAMES.contains(pkg.get_name())) {
+            filter_result.add_unsafe(pkg.get_id().id);
+        }
+    }
+
+    libdnf5::advisory::AdvisoryQuery advisories{p_impl->base};
+    auto reboot_advisories = advisories.get_advisory_packages_sorted_by_name_arch_evr();
+    std::erase_if(reboot_advisories, [](const auto & pkg) { return !pkg.get_reboot_suggested(); });
+
+    const auto & cmp_naevr = libdnf5::rpm::cmp_naevr<const advisory::AdvisoryPackage &, const Package &>;
+    for (const auto & pkg : *this) {
+        auto lower = std::lower_bound(reboot_advisories.begin(), reboot_advisories.end(), pkg, cmp_naevr);
+        if (lower != reboot_advisories.end() && lower->get_nevra() == pkg.get_nevra()) {
+            filter_result.add_unsafe(pkg.get_id().id);
+        }
+    }
+
+    *p_impl &= filter_result;
```

Comment:
Oh, this is my bad, I chose a wrong name for the `get_advisory_packages_sorted_by_name_arch_evr()` function. I really need to at least document it.

The problem is that it returns the advisory packages sorted by libsolv `id`s of name, arch and evr, not the actual strings they represent. It is a different sorting than `cmp_naevr` and therefore the `std::lower_bound(...)` doesn't work correctly.

Perhaps the best solution would be to replace this block with:
```cpp
PQImpl::filter_sorted_advisory_pkgs(*this, reboot_advisories, libdnf5::sack::QueryCmp::EQ);
```
(It should also be faster because it compares the `id`s directly and doesn't resolve them into strings.)

Reply:
I have created: https://github.com/rpm-software-management/dnf5/pull/998

Reply:
Ah thanks, I thought the results from `filter_reboot_suggested` looked correct, but you're right, it wasn't working. I think it's correct now, . Clearly some tests are needed still.